src/main/java/org/terasology/logic/health/RegenAuthoritySystem.java

```diff
@@ -95,39 +107,48 @@ private void invokeRegenOperations(long currentWorldTime) {
             }
         });
 
+        // Regenerate the entities with EndTime greater than Current time
         regenerate(currentWorldTime);
     }
 
     private void regenerate(long currentTime) {
-        for (EntityRef entity: regenSortedByTime.values()) {
+        List<Long> regenToBeRemoved = new LinkedList<>();
+        for (EntityRef entity : regenSortedByTime.values()) {
             RegenComponent regen = entity.getComponent(RegenComponent.class);
             HealthComponent health = entity.getComponent(HealthComponent.class);
             if (health.nextRegenTick < currentTime) {
                 logger.warn("regenerating " + regen.getRegenValue());
                 health.currentHealth += regen.getRegenValue();
                 health.nextRegenTick = currentTime + 1000;
                 if (health.currentHealth >= health.maxHealth) {
-                    regenSortedByTime.remove(regen.getLowestEndTime(), entity);
+                    regenToBeRemoved.add(regen.getLowestEndTime());
                     entity.removeComponent(RegenComponent.class);
                     entity.send(new OnFullyHealedEvent(entity));
                 }
                 entity.saveComponent(health);
             }
         }
+        for (Long endTime : regenToBeRemoved) {
```

Comment:
If you feel like being fancy you _can_ use a lambda here.
Not suggesting that you do do this, leaving it as a loop is clear. But just giving you more info.